Determine collection_type for all workflow steps

mvdbeek opened the pull request:
This allows us to get the correct collection type for subworkflows, which is a pretty major aspect if you want to use multiple subworkflows or want to link subworkflow outputs to collection inputs. This probably doesn't return the correct nesting level for unlinked, multiplied inputs (but that also doesn't work for regular workflows at this time, and I think it can't). But this already enables things like this:
![screen shot 2019-01-22 at 19 41 53](https://user-images.githubusercontent.com/6804901/51558715-75a67080-1e80-11e9-868a-e470042d113b.png)

mvdbeek committed:
Determine collection_type for all workflow steps

mvdbeek committed:
Unit test resolve_collection_type

mvdbeek committed:
Don't fail if tool is not installed

jmchilton commented:
Unit tests are failing on Travis. Looks really good though - this is obviously really needed and thanks for doing it.

mvdbeek commented:
Thanks @jmchilton! I noticed another small issue when actually using this, we should only fill in `collection_type` for subworkflows. If we do this for the parent we nest way too deep, because the client also attempts to figure out the collection type. And I guess this is the right behaviour.

mvdbeek committed:
Only fill collection_type for subworkflows, the top level stuff is managed by the client

mvdbeek committed:
Don't display subworkflow inputs as workflow_outputs

mvdbeek committed:
Add input collection_type when including subworkflow

mvdbeek commented:
5ac0802 fixes the subworkflow's input data collection appearing in the workflow outputs (see 0:output in the list:list input to list:list:list example in the PR description) and c5726c3 fills in the collection_type for collection input modules, which fixes the mapping over calculation.

I think this is ready again.

nsoranzo commented:
Unit tests are failing: https://travis-ci.org/galaxyproject/galaxy/jobs/485428749

mvdbeek commented:
Ah, that's probably relevant -- we don't force a `collection_type` to be set. I'll make this more defensive.

mvdbeek committed:
Only add collection_type if collection_type has been specified

We don't force the collection_type to be set in any way.
We might want to think about defaulting to list if not specified,
but that's not for now.
